rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala

```diff
@@ -0,0 +1,401 @@
+      * 3. Produce on the channel value:
+      *     - If a continuation exists at the channel,
+      *       match sender pars against abstraction patterns.
+      *       Pattern bindings are added to the sender environment.
+      *     - If a continuation does not exist at the channel,
+      *       return unit.
+      * 4. Merge the abstraction environment with the updated
+      *    sender environment.
+      * 5. Interpreter the body of the abstraction in the
+      *    new environment.
+      * @param send An output process
+      * @param env0 An execution context
+      * @return
+      */
+    def eval(send: Send)(implicit env: Env[Par]): Task[Unit] =
+      for {
+        quote           <- eval(send.chan.get)
+        data            <- send.data.toList.traverse(x => evalExpr(x)(env))
+        subChan: Quote  = substitute(quote)
+        optContinuation <- produce(subChan, data, send.persistent)
+
+        _ <- optContinuation match {
+              case Some((body: Par, newEnv: Env[Par])) =>
+                continue(body)(newEnv)
+              case None => Task.unit
+            }
+      } yield ()
+
+    def eval(receive: Receive)(implicit env: Env[Par]): Task[Unit] =
+      for {
+        binds <- receive.binds.toList
+                  .traverse((rb: ReceiveBind) =>
+                    eval(rb.source.get).map(quote => (rb.patterns, substitute(quote))))
+        optContinuation <- consume(binds, receive.body.get, receive.persistent)
+        _ <- optContinuation match {
+              case Some((body: Par, newEnv: Env[Par])) =>
+                continue(body)(newEnv)
+              case None => Task.unit
+            }
+      } yield ()
+
+    /**
+      * Eval is well-defined on channel variables provided
+      * a binding exists for the variable. It should
+      * always yield either a Par or an error.
+      */
+    def eval(drop: Eval)(implicit env: Env[Par]): Task[Unit] =
+      for {
+        quote <- eval(drop.channel.get)
+        _     <- eval(quote.value)
+      } yield ()
+
+    /**
+      * The constructor "now", used above in alloc,
+      * guarantees that no new thread is forked.
+      * The flatMap call sequences alloc and eval -
+      * so New executes sequentially.
+      *
+      * @param neu
+      * @param env
+      * @return
+      */
+    def eval(neu: New)(implicit env: Env[Par]): Task[Unit] = {
+      def alloc(level: Int)(implicit env: Env[Par]): Task[Env[Par]] =
+        Task now {
+          (Env[Par]() /: (0 to level).toList) { (_env, _) =>
+            val addr: Par = GPrivate()
+            _env.put(addr)
+          }
+        }
+      for {
+        _env <- alloc(neu.bindCount)
+        _    <- eval(neu.p.get)
+      } yield ()
+    }
+
+    def evalToInt(p: Par)(implicit env: Env[Par]): Task[Int] = {
+      if (!p.sends.isEmpty || !p.receives.isEmpty || !p.evals.isEmpty || !p.news.isEmpty || !p.matches.isEmpty || !p.ids.isEmpty)
```

Comment:
This if statement doesn't do anything. If there is indeed some other junk, it will just go onto `case _ => Task raiseError new Error("Error: Integer expected, or unimplemented expression.")`

Reply:
There should be an else. Oops.